eng/packages/http-client-csharp/generator/Azure.Generator/src/Providers/ResourceProvider.cs

```diff
@@ -0,0 +1,345 @@
+                    : operationVariable.Invoke("WaitForCompletion", KnownParameters.CancellationTokenParameter).Terminate()
+                });
+                tryStatement.Add(Return(operationVariable));
+            }
+            else
+            {
+                tryStatement.Add(new IfStatement(responseVariable.Property("Value").Equal(Null))
+            {
+                ((KeywordExpression)ThrowExpression(New.Instance(typeof(RequestFailedException), responseVariable.Invoke("GetRawResponse")))).Terminate()
+            });
+                tryStatement.Add(Return(Static(typeof(Response)).Invoke(nameof(Response.FromValue), New.Instance(Type, This.Property("Client"), responseVariable.Property("Value")), responseVariable.Invoke("GetRawResponse"))));
+            }
+            return tryStatement;
+        }
+
+        private static CSharpType GetResponseType(MethodProvider convenienceMethod, bool isAsync) => isAsync ? convenienceMethod.Signature.ReturnType?.Arguments[0]! : convenienceMethod.Signature.ReturnType!;
+
+        private ValueExpression[] PopulateArguments(IReadOnlyList<ParameterProvider> parameters)
+        {
+            var arguments = new List<ValueExpression>();
+            foreach (var parameter in parameters)
+            {
+                if (parameter.Name.Equals("subscriptionId", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    arguments.Add(Static(typeof(Guid)).Invoke(nameof(Guid.Parse), This.Property(nameof(ArmResource.Id)).Property(nameof(ResourceIdentifier.SubscriptionId))));
+                    continue;
+                }
+                if (parameter.Name.Equals("resourceGroupName", StringComparison.InvariantCultureIgnoreCase))
+                {
+                    arguments.Add(This.Property(nameof(ArmResource.Id)).Property(nameof(ResourceIdentifier.ResourceGroupName)));
+                    continue;
+                }
+
+                // TODO: handle parent
+                if (parameter.Name.Equals(_contextualParameters.Last(), StringComparison.InvariantCultureIgnoreCase))
+                {
+                    arguments.Add(This.Property(nameof(ArmResource.Id)).Property(nameof(ResourceIdentifier.Name)));
+                    continue;
+                }
+
+                arguments.Add(parameter);
+            }
+            return arguments.ToArray();
+        }
+
+        // TODO: get clean name of operation Name
+        private MethodProvider GetCorrespondingConvenienceMethod(InputOperation operation, bool isAsync)
+        {
+            var methods = _clientProvider.Methods;
+            return methods.Single(m => m.Signature.Name.Equals(isAsync ? $"{operation.Name}Async" : operation.Name, StringComparison.OrdinalIgnoreCase) && m.Signature.Parameters.Any(p => p.Type.Equals(typeof(CancellationToken))));
+        }
```

Comment:
considering the existence of customization code, should we get the methods from the canonical type of `_clientProvider`?
In MGC, when we override a method via customization code, the `_clientProvider` represents the "generated part" of that  class, it will not have that method any more. But the canonical type should always be its full view of the type.

Reply:
use _clientProvider.CutomCodeView for now, added TODO to use _clientProvider.CanonicalView